pkg/lvm/lvm_util.go

```diff
@@ -299,6 +333,45 @@ func CreateVolume(vol *apis.LVMVolume) error {
 		return nil
 	}
 
+	// Handle volume creation based on source
+	if vol.Spec.Source != "" {
+		// check if the source is snapshot
+		if !strings.HasPrefix(vol.Spec.Source, "snapshot-") {
+			return fmt.Errorf("unsupported source %s; only snapshot source is supported", vol.Spec.Source)
+		}
+
+		snapName := getLVMSnapName(vol.Spec.Source)
+		if exists, err := isSnapshotExists(vol.Spec.VolGroup, snapName); err != nil {
+			return fmt.Errorf("failed to check snapshot existence: %w", err)
+		} else if !exists {
+			return fmt.Errorf("snapshot %s does not exist in volume group %s", snapName, vol.Spec.VolGroup)
+		}
+
+		// Create volume from thin snapshot
+		args := buildLVMRestoreThinSnapshotArgs(vol)
+		out, _, err := RunCommandSplit(LVCreate, args...)
+		if err != nil {
+			err = newExecError(out, err)
+			klog.Errorf(
+				"lvm: could not create snapshot %v cmd %v error: %s", volume, args, string(out),
```

Comment:
Just to clearly differentiate that we are trying to create snapshot for restore purpose.
```suggestion
				"lvm: restore: could not create snapshot %v cmd %v error: %s", volume, args, string(out),
```

Reply:
[updated](https://github.com/openebs/lvm-localpv/pull/419/commits/a181feafcb90b8cc77ca96fdb9e75e8cb95aa55a)